DOC-4291 Rewards administration guide (#101)
exoplatform/exo-documentation: docs/Configuration.rst

-  The **EventCategoriesCache** size is approximately 24 bytes, so the
   maximum heap size equals to the cache size multiplied by 24 bytes.
   

Wallet caches
~~~~~~~~~~~~~

eXo Platform provides 2 wallet caches:

-  :ref:`Wallet account cache <Wallet.accountCache>`

-  :ref:`Wallet transaction cache <Wallet.transactionCache>`

You can add new values of these caches in :ref:`exo.properties <Configuration.ConfigurationOverview>` file.

.. code-block:: jproperties

    exo.cache.wallet.account.MaxNodes=2000
    exo.cache.wallet.account.TimeToLive=-1
    exo.cache.wallet.transactions.MaxNodes=2000
    exo.cache.wallet.transactions.TimeToLive=-1

The specific configuration of wallet caches can be found in:

-  ``$PLATFORM_TOMCAT_HOME/webapps/wallet-common.war!/WEB-INF/wallet/cache-configuration.xml``.

.. _Wallet.transactionCache:   

Wallet account cache
-------------------------

The **Wallet account cache** caches the Wallet objects.
This object contains metadata information of a wallet, such as identity id, username, space pretty name, wallet enablement...

> When any user get access to his wallet or a space wallet, the cached wallet object will be retrieved from cache rather than the database.

Reviewer: ```suggestion
When any user get access to their wallet or a space wallet, the cached wallet object will be retrieved from cache rather than the database.
```

Author: Done

> When any user get access to their wallet or a space wallet, the cached wallet object will be retrieved from cache rather than the database.

Reviewer 2: to their=>to his

Author: as Patrice said, we don't know if it's a he or she :-)
So we use their.
See https://jakubmarian.com/he-or-she-vs-they-in-english/

> Finally, it is becoming increasingly widespread and regarded as completely grammatically correct to use a plural pronoun for a singular noun to refer to it in a gender-neutral way

Reviewer: yep. tricky one but convenient and gender neutral.

-  The cached **Wallet account** is invalidated when the user updates its metadata such as reimporting its private key.

> -  The **Wallet account cache** size should equals the number of wallets used in eXo Platform.

Reviewer 2: equal (after should)

Author: done

> -  The **Wallet account cache** size should equals to the number of wallets used in eXo Platform.

Reviewer: ```suggestion
-  The **Wallet account cache** size should equals the number of wallets used in eXo Platform.
```

Author: Done

> -  Each Wallet object is approximately 400 bytes, so the maximum heap size equals to the cache size multiplied by 400 bytes.

Reviewer 2: Each Wallet object take about 400 bytes in memory. So the maximum heap size occupied by the wallet cache would be its size multiplied by 400 bytes

Author: done

.. _Wallet.transactionCache:   

Wallet Transactions cache
-------------------------

The **Wallet Transactions cache** caches the blockchain transaction metadata sent by a wallet. This object contains information such as transaction hash, username, space pretty name, transaction status...

> When any user get access to the list of his transactions or a space wallet transactions, the cached wallet transactions object will be retrieved from cache rather than the database.

Reviewer: ```suggestion
When any user get access to the list of their transactions or a space wallet transactions, the cached wallet transactions object will be retrieved from cache rather than the database.
```

Author: Done

> When any user get access to the list of their transactions or a space wallet transactions, the cached wallet transactions object will be retrieved from cache rather than the database.

Reviewer 2: their => his

Reviewer 2: object=> objects

Author: done for https://github.com/exoplatform/exo-documentation/pull/101#discussion_r310129425 only.

for the first remark, see https://github.com/exoplatform/exo-documentation/pull/101#discussion_r310144515

> -  The cached **Wallet transaction** is invalidated when the user updates its metadata such as changing transaction status from *pending* to *success* or *failed*.

Reviewer: ```suggestion
-  The cached **Wallet transaction** is invalidated when the users updates their metadata such as changing transaction status from *pending* to *success* or *failed*.
```

Author: Done

> -  The **Wallet transaction cache** size should equals to the number of wallet account multiplied by the the first transactions list page size (10 elements per page).

Reviewer: ```suggestion
-  The **Wallet transaction cache** size should equals the number of wallet accounts multiplied by the first transactions list page size (10 elements per page).
```

-  Each Wallet transaction object is approximately 700 bytes, so the maximum heap size equals to the cache size multiplied by 700 bytes.

> 

Reviewer 2: Each Wallet transaction object take about 700 bytes in memory. So the maximum heap size occupied by the wallet transaction cache would be its size multiplied by 700 bytes

Author: done

      The periodicity can be configured using the following property (Default value: every hour):

    .. code-block:: jproperties

          exo.wallet.ContractTransactionVerifierJob.expression=0 0 * ? * * *

   -  Checks all pending transactions sent using wallet application to update its state internally in internal store and to send notifications for receiver and sender.
      The check is made periodically and can be configured using the following property (Default value: every 10 seconds):

    .. code-block:: jproperties

          exo.wallet.PendingTransactionVerifierJob.expression=0/10 * * * * ?

   -  Send transactions on blockchain using *Admin wallet* that is used to administrate all other wallets.
      In fact this wallet is used to initialize and send rewards to users and spaces wallets.

- Unlike the *Admin wallet* which uses `Web3j <https://web3j.io/>`__ (Server side) to issue transactions, the HTTP endpoint is used by *users and spaces wallets* using `Web3.js <https://web3js.readthedocs.io/>`__ (Browser side) to communicate with the blockchain to issue transactions.

> // TODO add blockchain chart, wallet UI and eXo server behind that communicates together (ASK DESIGNER)

Reviewer 3: 👍

      -  Cheap transaction: this option will use the gas price configured by the following property (in `WEI <http://ethdocs.org/en/latest/ether.html>`__):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.cheapPrice=4000000000

      -  Normal transaction: this option will use the gas price configured by the following property (in `WEI <http://ethdocs.org/en/latest/ether.html>`__):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.normalPrice=8000000000

      -  Fast transaction: this option will use the gas price configured by the following property (in `WEI <http://ethdocs.org/en/latest/ether.html>`__):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.fastPrice=15000000000

* When a transaction is sent and wasn't mined for several days, it will be marked as failed in eXo Wallet internal database.
  To configure the number of days waiting for transaction to get mined, please use the following property:

    .. code-block:: jproperties

          exo.wallet.transaction.pending.maxDays=3

> 

Reviewer 3: It it possible to the transaction to be mined after this delay ? What will happen to its status on the internal database ?

Author: Exactly, I was thinking to add the detail here, because it happened before.

Author: Added and described here:
https://github.com/exoplatform/exo-documentation/pull/101/files#diff-617637c973b50abec045ab9fee0de609R6125

To disable the login history data storage in the database, simply set the parameter 
``exo.audit.login.enabled`` to ``false``.

.. _Configuration.Wallet:

================
eXo Wallet addon
================

> The eXo Wallet addon use `Ethereum Blockchain <https://www.ethereum.org/>`__ to reward users based on a crypto-currency (*Cauri*) created inside this blockchain.

Reviewer: ```suggestion
The eXo Wallet addon uses the `Ethereum Blockchain <https://www.ethereum.org/>`__ to reward users based on a crypto-currency managed inside this blockchain.
```

Author: Done

    exo.wallet.blockchain.network.websocket=wss://mainnet.infura.io/ws/v3/a1ac85aea9ce4be88e9e87dad7c01d40
    exo.wallet.blockchain.network.http=https://mainnet.infura.io/v3/a1ac85aea9ce4be88e9e87dad7c01d40

- The Websocket endpoint is used to:

   -  Scan new mined blocks on blockchain to verify if there is a new mined transaction that involves a wallet of a recognized user.
      The periodicity can be configured using the following property (Default value: every hour):

    .. code-block:: jproperties

          exo.wallet.ContractTransactionVerifierJob.expression=0 0 * ? * * *

   -  Checks all pending transactions sent using wallet application to update its state internally in internal store and to send notifications for receiver and sender.
      The check is made periodically and can be configured using the following property (Default value: every 10 seconds):

    .. code-block:: jproperties

          exo.wallet.PendingTransactionVerifierJob.expression=0/10 * * * * ?

   -  Send transactions on blockchain using *Admin wallet* that is used to administrate all other wallets.
      In fact this wallet is used to initialize and send rewards to users and spaces wallets.

> - The HTTP endpoint is used by *users and spaces wallets* using `Web3.js <https://web3js.readthedocs.io/>`__ (Browser side) to communicate with the blockchain to send transactions, unlike the *Admin wallet* which uses `Web3j <https://web3j.io/>`__ (Server side) to send transactions.

Reviewer: ```suggestion
- Unlike the *Admin wallet* which uses `Web3j <https://web3j.io/>`__ (Server side) to issue transactions, the HTTP endpoint is used by *users and spaces wallets* using `Web3.js <https://web3js.readthedocs.io/>`__ (Browser side) to communicate with the blockchain to issue transactions.
```

Author: Done

> // TODO ajouter un schéma avec la blockchain, la UI d'eXo Wallet et le server eXo derrière qui communiquent ensemble (A DEMANDER AU DESIGNER de l'aide !!!!)

Reviewer: ```suggestion
// TODO add blockchain chart, wallet UI and eXo server behind that communicates together (ASK DESIGNER)
```
english please

Author: Done

Blockchain transaction settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> - When sending a transaction to Ethereum blockchain, some generic properties are required, especially *gas limit* and *gas price*.

Reviewer: ```suggestion
- When sending a transaction to the Ethereum blockchain, some generic properties are required, especially *gas limit* and *gas price*.
```

Author: Done

  - **Gas limit**

>     The *gas limit* is the maximum gas that can take a transaction using *Cauri* contract. It can be modified using the following property:

Reviewer: ```suggestion
    The *gas limit* is the maximum gas that can take a transaction using the token contract. It can be modified using the following property:
```
perhaps we could have an external link to a definition of the term gas

Author: Done

      .. code-block:: jproperties

            exo.wallet.transaction.gas.limit=150000

  - **Gas price**

    The *gas price* parameters will determine transaction fee of operations made using eXo Wallet addon. In the UI, the user has three choices:

>       -  Cheap transaction: this option will use the gas price configured by the following property (in WEI):

Reviewer: external link to definition of WEI ?

Author: Done

      .. code-block:: jproperties

            exo.wallet.transaction.gas.cheapPrice=4000000000

      -  Normal transaction: this option will use the gas price configured by the following property (in WEI):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.normalPrice=8000000000

      -  Fast transaction: this option will use the gas price configured by the following property (in WEI):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.fastPrice=15000000000

* When a transaction is sent and wasn't mined for several days, it will be marked as failed in eXo Wallet internal database.
  To configure the number of days waiting for transaction to get mined, please use the following property:

    .. code-block:: jproperties

          exo.wallet.transaction.pending.maxDays=3

Wallet types
~~~~~~~~~~~~

* **Admin wallet**

>   The *Admin wallet* is created automatically during first startup of eXo Platform server.

Reviewer: ```suggestion
  The *Admin wallet* is created automatically during first startup of the eXo Platform server.
```

Author: Done

>   In order to be able to send transactions on blockchain using `Web3j<https://web3j.io/>` in server side, the private key of the *Admin wallet* is stored in internal database.

Reviewer: ```suggestion
  In order to be able to issue transactions on the blockchain using `Web3j<https://web3j.io/>` server side, the private key of the *Admin wallet* is stored (in encrypted form) in the internal database.
```

Author: Done

>   The private key is encrypted using two passwords:

Reviewer: ```suggestion
  The private key is encrypted by combining two passwords:
```

Author: Done

>      -  One password that is coming from a keystore file ( see :ref:`Updating password encryption key <UpdatePasswordEncryptionKey>` )

Reviewer: ```suggestion
     -  A first password is read from a keystore file ( see :ref:`Updating password encryption key <UpdatePasswordEncryptionKey>` )
```

Author: Done

>      .. warning:: The generated keystore file under `gatein/conf/codec` has to be backed up as a data folder, because it have a key file that is used to decrypt stored wallets private keys. If it's lost, all wallets private keys will be lost and consequently, all funds would be lost as well.

Reviewer: ```suggestion
     .. warning:: The generated keystore file under `gatein/conf/codec` MUST be backed up as a data folder, because it contains a key file that is used to decrypt stored wallets private keys. If it's lost, all wallets private keys will be lost and consequently, all funds would be lost and unrecoverable.
```

>      -  And another password that can be configured in properties:

Reviewer: ```suggestion
     -  A second password that can be configured in properties:
```

    .. code-block:: jproperties

          exo.wallet.admin.key=changeThisKey

>     .. warning:: The password couldn't be changed once the platform is started for the first time. In fact, this password will be used to encrypt *Admin wallet* private key that will be stored in database.

Reviewer: ```suggestion
    .. warning:: The password can't be changed once the platform is started for the first time. In fact, this password will be used to encrypt the *Admin wallet*'s private key that will be stored in database.
```

>         If its value is modified after server startup, the private key of admin wallet couldn't be decrypted anymore.

Reviewer: ```suggestion
        If its value is modified after server startup, the private key of admin wallet won't be decrypted anymore, preventing all administrative operations.
```

  The *Admin wallet* can be used by */platform/rewarding* group members to initialize other wallets.

>   The private key of the admin wallet cannot be accessed by any user to avoid sharing its funds with a person.

Reviewer: ```suggestion
  The private key of the admin wallet cannot be accessed by any user to avoid exposing its funds to an unauthorized person.
```

* **User wallet**

  By default, any user of the platform can create his own wallet. A property is added to restrict access to wallet application to a group of users:

    .. code-block:: jproperties

          # Example of expressions
          
          # Any user
          exo.wallet.accessPermission=
          # All connected users of the platform
          exo.wallet.accessPermission=/platform/users
          # All administrators of the platform
          exo.wallet.accessPermission=*:/platform/administrators
          # All members of space associated to group internal_space
          exo.wallet.accessPermission=member:/spaces/internal_space

* **Space wallet**

>   Space managers can add wallet application and associate a wallet for a space using :ref:`Managing space applications <ManagingSpaceApplication>` UI:

Reviewer: ```suggestion
  Space managers can add the wallet application in their space and associate a wallet for a space using :ref:`Managing space applications <ManagingSpaceApplication>` UI:
```

  |SpaceAddWallet|

>   Space wallets can be managed by one or multiple space manager. In fact, the space wallet is similar to user wallet, to be able to activate it, the *Admin wallet* has to initialize it.

Reviewer: ```suggestion
  Space wallets can be managed by one or multiple space manager. In fact, the space wallet is similar to user wallet. To use it, it must be activated by an admin. The operation is carried out by the *Admin Wallet* account.
```

  By default, any space member can access to the space wallet in *readonly* mode.

> Cauri contract

Reviewer: ```suggestion
Token contract
```

~~~~~~~~~~~~~~

> The used Cauri contract address can be configured using the following variable:

Reviewer: ```suggestion
The contract address used for the main token currency can be configured using the following variable:
```

    .. code-block:: jproperties

          exo.wallet.blockchain.token.address=0xc76987d43b77c45d51653b6eb110b9174acce8fb

> This address will never change through eXo Platform versions on Ethereum Mainnet blockchain and is shared with all eXo Community.

Reviewer: ```suggestion
This address is the official contract address for the official rewarding token. It shouldn't change through eXo Platform versions on the Ethereum Mainnet blockchain. This crypto-currency token, is shared by all eXo Community users.
```

> The *Cauri* contract is an `ERC-20 <https://eips.ethereum.org/EIPS/eip-20>`__ Token contract that is deployed on Ethereum Mainnet.

Reviewer: ```suggestion
It is an `ERC-20 <https://eips.ethereum.org/EIPS/eip-20>`__ Token contract that is deployed on Ethereum Mainnet.
```

> To be able to receive some *Cauri*, a wallet has to be initialized by a *Cauri* contract administrator on blockchain (not on eXo Platform Server bundle that you own).

Reviewer: ```suggestion
To be able to receive some tokens, a wallet has to be initialized by a contract administrator on the blockchain (this cannot be done on your eXo Platform server).
```

> // TODO ajouter un schéma avec PLUSIEURS exo déployés et au centre la blockchain où on trouve le CONTRAT CAURI (A DEMANDER AU DESIGNER de t'aider !!!!)

Reviewer: ```suggestion
// TODO add a diagram with many eXo instance deployed and in the center, the blockchain where the token contract lies (ASK DESIGNER)
```
english please

To disable the login history data storage in the database, simply set the parameter 
``exo.audit.login.enabled`` to ``false``.

.. _Configuration.Wallet:

======================
Configuring eXo Wallet
======================

> The eXo Wallet addon uses the `Ethereum Blockchain <https://www.ethereum.org/>`__ to reward users based on a crypto-currency managed inside this blockchain.

Reviewer: ```suggestion
The eXo Wallet addon uses the `Ethereum Blockchain <https://www.ethereum.org/>`__ to let eXo users hold and exchange crypto-currency (aka tokens) managed inside this blockchain.
```

Author: done

Blockchain network settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The addon is configured by default to communicate with Ethereum Mainnet throught `Infura <https://infura.io/>`__ using websocket and http protocols:

Reviewer: ```suggestion
The addon is configured by default to communicate with Ethereum's public network (called the "Mainnet") throught `Infura <https://infura.io/>`__ using websocket and http protocols:
```

Author: done

.. code-block:: jproperties

    exo.wallet.blockchain.networkId=1
    exo.wallet.blockchain.network.websocket=wss://mainnet.infura.io/ws/v3/a1ac85aea9ce4be88e9e87dad7c01d40

>     exo.wallet.blockchain.network.http=https://mainnet.infura.io/v3/a1ac85aea9ce4be88e9e87dad7c01d40

Reviewer: maybe explain that the default configuration points to an eXo-owner infura account that serves as proxy to the mainnet.

Author: done

- The Websocket endpoint is used to:

>    -  Scan new mined blocks on blockchain to verify if there is a new mined transaction that involves a wallet of a recognized user.

Reviewer: ```suggestion
   -  Scan newly mined blocks on the blockchain to check if any validated transaction involves the wallet owned by a user of your eXo Platform instance.
```

Author: done

      The periodicity can be configured using the following property (Default value: every hour):

    .. code-block:: jproperties

          exo.wallet.ContractTransactionVerifierJob.expression=0 0 * ? * * *

>    -  Checks all pending transactions sent using wallet application to update its state internally in internal store and to send notifications for receiver and sender.

Reviewer: ```suggestion
   -  Checks all pending transactions sent using the eXo Wallet application to update its internal state (cache and persistent storage) and send notifications to receiver and sender of the transactions.
```

Author: done

      The check is made periodically and can be configured using the following property (Default value: every 10 seconds):

    .. code-block:: jproperties

          exo.wallet.PendingTransactionVerifierJob.expression=0/10 * * * * ?

>    -  Send transactions on blockchain using *Admin wallet* that is used to administrate all other wallets.

Reviewer: ```suggestion
   -  Send transactions from the so-called *Admin wallet* which is a dedicated blockchain account with special privileges than is used to administrate (initialize, approved, delete, reward, etc.)  all user wallets.
```

Author: In fact it initializes and rewards users only

Author: done

>       In fact this wallet is used to initialize and send rewards to users and spaces wallets.

Reviewer: ```suggestion
```
included in previous suggestion

Author: done

- Unlike the *Admin wallet* which uses `Web3j <https://web3j.io/>`__ (Server side) to issue transactions, the HTTP endpoint is used by *users and spaces wallets* using `Web3.js <https://web3js.readthedocs.io/>`__ (Browser side) to communicate with the blockchain to issue transactions.

  |WalletBlockchainCommunication|

Blockchain transaction settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> - When sending a transaction to the Ethereum blockchain, some generic properties are required, especially *gas limit* and *gas price*.

Reviewer: ```suggestion
- When submitting a transaction to the Ethereum blockchain, some special properties are required by the blockchain protocols, namely *gas limit* and *gas price*.
```

Author: done

  - **Gas limit**

>     The *gas limit* is the maximum gas that can take a transaction using the token contract. It can be modified using the following property:

Reviewer: ```suggestion
    The *gas limit* is the maximum gas that a transaction can consume to execute any operation using the token contract. It can be modified using the following property:
```

Author: done

      .. code-block:: jproperties

>             exo.wallet.transaction.gas.limit=150000

Reviewer: the description does not tell the unit of this property. gwei ?

Author: Sorry, for the gas limit, it measures the complexity of a transaction, and not a price. It's unit it's "gas". So if I read this property content, I should tell "gas limit equals to 150000 gas"

Author: done

  - **Gas price**

    The *gas price* parameters will determine transaction fee of operations made using eXo Wallet addon. In the UI, the user has three choices:

      -  Cheap transaction: this option will use the gas price configured by the following property (in `WEI <http://ethdocs.org/en/latest/ether.html>`__):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.cheapPrice=4000000000

      -  Normal transaction: this option will use the gas price configured by the following property (in `WEI <http://ethdocs.org/en/latest/ether.html>`__):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.normalPrice=8000000000

      -  Fast transaction: this option will use the gas price configured by the following property (in `WEI <http://ethdocs.org/en/latest/ether.html>`__):

      .. code-block:: jproperties

            exo.wallet.transaction.gas.fastPrice=15000000000

> * When a transaction is sent and wasn't mined for several days, it will be marked as failed in eXo Wallet internal database.

Reviewer: ```suggestion
* When a transaction is submitted but isn't mined for several days, it will be marked as failed in eXo Wallet's internal database.
```

Author: done

  The maximum number of days waiting for transaction mining can be configured using the following property:

    .. code-block:: jproperties

          exo.wallet.transaction.pending.maxDays=3

>     .. note:: If the transaction still not mined after ``exo.wallet.transaction.pending.maxDays`` days, it will be marked as failed transaction in internal eXo database.

Reviewer: ```suggestion
    .. note:: If the transaction is still not mined after ``exo.wallet.transaction.pending.maxDays`` days, it will be marked as a failed transaction in eXo's internal database.
```

>               When it gets mined, the Job ``ContractTransactionVerifierJob`` will detect it and will store again the real transaction status coming from blockchain and send notifications.

Reviewer: ```suggestion
              if/when it gets eventually mined (pending.maxDays), the Job ``ContractTransactionVerifierJob`` will detect it and will update the real transaction status coming from the blockchain and send notifications.
```

Wallet types
~~~~~~~~~~~~

* **Admin wallet**

  The *Admin wallet* is created automatically during first startup of the eXo Platform server.

>   In order to be able to issue transactions on the blockchain using `Web3j<https://web3j.io/>`__ server side, the private key of the *Admin wallet* is stored (in encrypted form) in the internal database.

Reviewer: ```suggestion
  In order to be able to submit transactions to the blockchain using `Web3j<https://web3j.io/>`__ server side, the private key of the *Admin wallet* is stored (in encrypted form) in the internal database.
```

  The private key is encrypted by combining two passwords:

     -  A first password is read from a keystore file ( see :ref:`Updating password encryption key <UpdatePasswordEncryptionKey>` )

     .. warning:: The generated keystore file under `gatein/conf/codec` MUST be backed up as a data folder, because it contains a key file that is used to decrypt stored wallets private keys. If it's lost, all wallets private keys will be lost and consequently, all funds would be lost and unrecoverable.

>      -  A second password that can be configured in properties:

Reviewer: ```suggestion
     -  A second password that can be configured in properties:
```

'can' suggestes that it is optional. Is it optional ? if not, use "must"

    .. code-block:: jproperties

          exo.wallet.admin.key=changeThisKey

    .. warning:: The password can't be changed once the platform is started for the first time. In fact, this password will be used to encrypt the *Admin wallet*'s private key that will be stored in database.

>         If its value is modified after server startup, the private key of admin wallet won't be decrypted anymore, preventing all administrative operations.

Reviewer: If an admin does this mistake (modifies after first startup), is the operation destrictive or a simple roll back to the previous value will let the admin operations work again ?

Also : is there a way to deliberately reset (e.g in case of compromision) the admin wallet private key ? Can this be done just by wiping out both  the key and the key store ?

Author: That's right, we must give the possibility to change the password. It can be done using a UI that I deleted (Admin account in Wallet admin). So unfortunately, for now, we don't have a solution to reset this password. (May be to add in the list of User stories of Feature, WDYT ?)

> If an admin does this mistake (modifies after first startup), is the operation destrictive or a simple roll back to the previous value will let the admin operations work again ?

Yep, it works again without affecting any transaction already sent, neither the already stored data in database.

  The *Admin wallet* can be used by */platform/rewarding* group members to initialize other wallets.
  The private key of the admin wallet cannot be accessed by any user to avoid exposing its funds to an unauthorized person.

* **User wallet**

>   By default, any user of the platform can create his own wallet. A property is added to restrict access to wallet application to a group of users:

Reviewer: ```suggestion
  By default, any user of the platform can create their own wallet. A property is added to restrict access to wallet application to a group of users:
```

the user can be a girl so "his" is not correct. any + their can be combined for ungendered meaning.

Author: right :-)

    .. code-block:: jproperties

          # Example of expressions
          
          # Any user
          exo.wallet.accessPermission=
          # All connected users of the platform
          exo.wallet.accessPermission=/platform/users
          # All administrators of the platform
          exo.wallet.accessPermission=*:/platform/administrators
          # All members of space associated to group internal_space
          exo.wallet.accessPermission=member:/spaces/internal_space

* **Space wallet**

>   Space managers can add the wallet application in their space a wallet for a space using :ref:`Managing space applications <ManagingSpaceApplication>` UI:

Reviewer: ```suggestion
  Space managers can add the wallet application in their space using :ref:`Managing space applications <ManagingSpaceApplication>` UI:
```

  |SpaceAddWallet|

>   Space wallets can be managed by one or multiple space manager. In fact, the space wallet is similar to user wallet. To use it, it must be activated by an admin. The operation is carried out by the *Admin Wallet* account.

Reviewer: ```suggestion
  Space wallets can be managed by one or multiple space manager. In fact, the space wallet is similar to user wallet. To use it, it must be activated by an admin. The operation is carried out by the *Admin Wallet* account.
```

it is not clear if space managers have different privileges than regular space members. Can all space managers send space wallet funds ?

Author: Space wallets can be managed by one or several space manager. In fact, by default the space wallet private key is stored in eXo internal database. So all managers has access to it and so they will automatically download it from server. But the the private key still encrypted using a password, that only wallet creator (that is one of space managers) owns it and could share with other managers to "send" funds from space wallet.

This description is long and may be, should be described more in User Guide (IMO) with more UI images.

  By default, any space member can access to the space wallet in *readonly* mode.

Token contract
~~~~~~~~~~~~~~

The contract address used for the main token currency can be configured using the following variable:

    .. code-block:: jproperties

          exo.wallet.blockchain.token.address=0xc76987d43b77c45d51653b6eb110b9174acce8fb

> This address is the official contract address for the official rewarding token. It shouldn't change through eXo Platform versions on the Ethereum Mainnet blockchain. This crypto-currency token, is shared by all eXo Community users.

Reviewer: ```suggestion
This address is the official contract address for the official rewarding token promoted by eXo. It shouldn't change through eXo Platform versions on the Ethereum Mainnet blockchain. This crypto-currency token, is shared by all eXo Community users.
```

It is an `ERC-20 <https://eips.ethereum.org/EIPS/eip-20>`__ Token contract that is deployed on Ethereum Mainnet.

> To be able to receive some tokens, a wallet address must be initialized by a *token contract* administrator on the blockchain (this cannot be done on your eXo Platform server).

Reviewer: ```suggestion
To be able to receive some tokens, a wallet address must be initialized by a *token contract* administrator (eXo staff) on the blockchain (this cannot be done on your eXo Platform server). 
```

Author: Not really *eXo staff* only. In fact, we have to promote *Admin wallet*, of eXo Platform instances deployed in customers/community servers, as administrator with level 2 on servers to be able to reward and initialize wallets of other users of their server instance.

======================
Configuring eXo Wallet
======================

The eXo Wallet addon uses the `Ethereum Blockchain <https://www.ethereum.org/>`__ to let eXo users hold and exchange crypto-currency (aka tokens) managed inside this blockchain.

Blockchain network settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~

The addon is configured by default to communicate with Ethereum's public network (called the "Mainnet") throught `Infura <https://infura.io/>`__ using websocket and http protocols:

.. code-block:: jproperties

    exo.wallet.blockchain.networkId=1
    exo.wallet.blockchain.network.websocket=wss://mainnet.infura.io/ws/v3/a1ac85aea9ce4be88e9e87dad7c01d40
    exo.wallet.blockchain.network.http=https://mainnet.infura.io/v3/a1ac85aea9ce4be88e9e87dad7c01d40

.. note:: 

    The default configuration points to an eXo-owner infura project ID that serves as proxy to the mainnet.
    To learn more about infura infrastructure and projects, please visit `Infura documentation <https://infura.io/docs/>`__

- The Websocket endpoint is used to:

   -  Scan newly mined blocks on the blockchain to check if any validated transaction involves a wallet owned by a user of your eXo Platform instance.

>       The periodicity can be configured using the following property (Default value: every hour):

Reviewer 2: of your eXo Platform instance => of your eXo Platform instance is available

Author: changed to:

`owned by an available user in your eXo Platform instance.`

Author: done

     -  A first password is read from a keystore file ( see :ref:`Updating password encryption key <UpdatePasswordEncryptionKey>` )

     .. warning:: The generated keystore file under `gatein/conf/codec` MUST be backed up as a data folder, because it contains a key file that is used to decrypt stored wallets private keys. If it's lost, all wallets private keys will be lost and consequently, all funds would be lost and unrecoverable.

     -  A second password that must be configured in properties:

    .. code-block:: jproperties

          exo.wallet.admin.key=changeThisKey

    .. warning:: The password can't be changed once the platform is started for the first time. In fact, this password will be used to encrypt the *Admin wallet*'s private key that will be stored in database.
        If its value is modified after server startup, the private key of admin wallet won't be decrypted anymore, preventing all administrative operations.

  The *Admin wallet* can be used by */platform/rewarding* group members to initialize other wallets.
  The private key of the admin wallet cannot be accessed by any user to avoid exposing its funds to an unauthorized person.

* **User wallet**

>   By default, any user of the platform can create their own wallet. A property is added to restrict access to wallet application to a group of users:

Reviewer 2: their=>his

Author: same remark as https://github.com/exoplatform/exo-documentation/pull/101#discussion_r310144515

    .. code-block:: jproperties

          # Example of expressions
          
          # Any user
          exo.wallet.accessPermission=
          # All connected users of the platform
          exo.wallet.accessPermission=/platform/users
          # All administrators of the platform
          exo.wallet.accessPermission=*:/platform/administrators
          # All members of space associated to group internal_space
          exo.wallet.accessPermission=member:/spaces/internal_space

* **Space wallet**

  Space managers can add the wallet application in their space using :ref:`Managing space applications <ManagingSpaceApplication>` UI:

  |SpaceAddWallet|

>   Space wallets can be managed by one or multiple space manager. In fact, the space wallet is similar to user wallet. To use it, it must be activated by an admin. The operation is carried out by the *Admin Wallet* account.

Reviewer 2: space manager=> space managers

Author: done
